allennlp/data/fields/field.py

```diff
@@ -116,7 +118,10 @@ def batch_tensors(self, tensor_list: List[DataArray]) -> DataArray:  # type: ign
 
     def __eq__(self, other) -> bool:
         if isinstance(self, other.__class__):
-            return self.__dict__ == other.__dict__
+            for attr in self.__slots__:
```

Comment:
What happens if we have a `Field` subclass the adds a new slot?  This would break, right?  I don't think we currently have this case, but it's something to think about.  I guess that class would have to also override `__eq__`?

Reply:
This works fine for subclasses without overriding `__eq__`. And I believe we have tests that cover this (see the `IndexField` tests, for example).

Reply:
I was imagining something like this:

```python
class Field:
    __slots__ = []

class IndexField(Field):
    __slots__ = ["index"]

class SpecialIndexField(IndexField):
    __slots__ = ["new_member_variable"]
```

From what I saw when I was reading about inheritance in slot classes, I'm pretty sure this would break without also overriding `__eq__`.  We don't have any such case right now, I was just wondering if this logic was actually safe in general.

Reply:
That should still work, but I'll add a test to make sure.

Reply:
Oh, you're saying that `self.__slots__` gets updated in the right way?  I'd be surprised, but I guess all of the slots stuff is magic to me, so maybe there's more magic I didn't know about.

Reply:
Ah, you're absolutely right. That does break it. I'm working on a fix.

Reply:
I mean, we have other cases where we have to override `__eq__`.  It's not the end of the world if it's impossible to make general logic here.  We should at least leave a note about the potential issue somewhere.

Reply:
Ok just added a test like that with a fix.

Reply:
When I run this locally, I get to `object` in `__bases__`, and get an attribute error because `object` doesn't have `__slots__`.

Reply:
Oh, hmm, are you using Python 2.*?

Reply:
3.7.  Not sure what the deal is.

Reply:
I'm not sure either, but try again now. The last commit should hopefully fix that for you

Reply:
I was just playing around in an interpreter because I was curious about what happened if there was a class in the chain that didn't have slots.  I was wondering how your tests were passing given what I saw.  But the logic that's there now would definitely work.

Reply:
I'm still not sure how your tests passed given what I saw in the interpreter, but 🤷‍♂️